ins-node/src/wifi_module/wifi_module.ino

```diff
@@ -1,10 +1,156 @@
+   @return            A tuple of (MAC address, RSSI)
+*/
+std::pair <String, int32_t> getNetworkContext(int8_t networkItem) {
+  std::pair <String, int32_t> context (wifi.BSSIDstr(networkItem), wifi.RSSI(networkItem));
+  return context;
+}
+
+/**
+   Measures the RSSI values for the various access points of the specified SSID
+   and returns them along the MAC address of the respective access point.
+   @return  The MAC addresses with their respective RSSI values
+*/
+std::vector<std::pair <String, int32_t>> getDatapoints() {
+  std::vector<std::pair <String, int32_t>> datapoints;
+  // Scan the surrounding networks and get how many were found
+  auto networksFound = wifi.scanNetworks();
+  for (auto i = 0; i < networksFound; ++i) {
+    if (wifi.SSID(i) == positioningSSID) {
+      datapoints.push_back(getNetworkContext(i));
+    }
+  }
+
+  return datapoints;
+}
+
+/**
+   Transmits the supplied data (if present) to the server
+   @param  datapoints The MAC addresses and the RSSI values we collected
+   @return            Whether transmission succesfull
+*/
+bool transmitData(std::vector<std::pair <String, int32_t>> datapoints) {
+  // TO-DO: Notify the server that we could not gather any datapoints for
+  // corrective actions to be made before sleeping.
+  if (datapoints.empty()) {
+    return false;
+  }
+  // Establish a TCP connection to the server
+  WiFiClient client;
+  if (!client.connect(SERVER_IP, HTTP_PORT)) {
+    return false;
+  }
+
+  // Compose a request for the server according to the agreed format:
+  // /set_rssi/device_id/mac_addr1/rssi1/mac_addr2/ ... /mac_addr10/:rssi10/
+  String request = "/set_rssi/";
+  request += DEVICE_ID;
+  request += "/";
+  for (auto datapoint : datapoints) {
+    // We don't put everything in the same expression since std::string
+    // (used in testing) is not exactly the same as Arduino String. Therefore
+    // we go with this way of concatinating the string which is compatible with
+    // both implementations.
+    request += datapoint.first; // The MAC address
+    request += "/";
+    request += datapoint.second; // The RSSI value
+    request += "/";
+  }
+
+  // Send the TCP request to the rest server
+  client.print("POST " + request + " HTTP/1.1\r\n" +
+               "Host: " + SERVER_IP + "\r\n" +
+               "Connection: close\r\n\r\n");
+
+  // Verify whether the request timed out
+  bool successfulRequest = true;
+  const unsigned long TIMEOUT = millis() + 5000;
+  while (client.available() == 0) {
+    if (millis() > TIMEOUT) {
+      successfulRequest = false;
+      break;
+    }
+  }
+  client.stop();
+  return successfulRequest;
+}
+
+/**
+   Puts the module into deep sleep and signals the power controller to
+   turn its power off.
+*/
+void goToSleep() {
+  digitalWrite(TX_PIN, HIGH);
+  delayMicroseconds(10);
+  digitalWrite(TX_PIN, LOW);
+  espPower.deepSleep(60 * 1000000UL); // Will sleep forever unless wakeup pin connected to RST
```

Comment:
Assert(Forever, 60 * 1000000UL) :smile:  I am curious how long this is.

Reply:
This is actually just 60 seconds as the input is in microseconds. However, since we have not connected a specific pin to the reset pin, the module will never manage to wake up after those 60 seconds. :smile: 
To be fair, the result would be the same if we would put `espPower.deepSleep(1)` instead. :stuck_out_tongue:

Reply:
Haha, I knew there was some magic stuffs going on there :laughing: